core/modules/xrdsvc/SsiRequest.cc

```diff
@@ -385,22 +387,33 @@ bool SsiRequest::replyFile(int fd, long long fSize) {
 bool SsiRequest::replyStream(StreamBuffer::Ptr const& sBuf, bool last) {
     // Create a streaming object if not already created.
     LOGS(_log, LOG_LVL_DEBUG, "replyStream, checking stream size=" << sBuf->getSize() << " last=" << last);
+
+    // Normally, XrdSsi would call Recycle() when it is done with sBuf, but if this function
+    // returns false, then it must call Recycle(). Otherwise, the scheduler will likely
+    // wedge waiting for the buffer to be released.
+    std::lock_guard<std::mutex> finLock(_finMutex);
+    if (_finished) {
+        // Finished() was called, give up.
+        sBuf->Recycle();
+        return false;
+    }
     if (!_stream) {
-       _stream = new ChannelStream();
-       if (SetResponse(_stream) != XrdSsiResponder::Status::wasPosted) {
-           LOGS(_log, LOG_LVL_WARN, "SetResponse stream failed, calling Recycle for sBuf");
-           // Normally, XrdSsi would call Recycle() when it is done with sBuf, but the
-           // return value from SetResponse indicates XrdSsi will never use sBuf nor call Recycle().
-           // Calling Recycle() here means we're done waiting for XrdSsi and sBuf can be freed when
-           // qserv is done with it.
-           sBuf->Recycle();
-           return false;
-       }
+        _stream = std::make_shared<ChannelStream>();
+        if (SetResponse(_stream.get()) != XrdSsiResponder::Status::wasPosted) {
+            LOGS(_log, LOG_LVL_WARN, "SetResponse stream failed, calling Recycle for sBuf");
+            // SetResponse return value indicates XrdSsi wont call Recycle().
+            sBuf->Recycle();
+            return false;
+        }
     } else if (_stream->closed()) {
+        // XrdSsi isn't going to call Recycle if we wind up here.
```

Comment:
This doesn't seem right. The stream gets closed it the last buffer has been added to the vector of buffers. If so, why is the code trying to add another one? That would indicate a serious logic error.

Reply:
I can add an error message to that affect, or throw. Either way, if this path is taken, Recycle should still be called as finding the source of a schedule wedge is difficult.